src/pip/_internal/commands/install.py

```diff
@@ -546,63 +536,83 @@ def site_packages_writable(**kwargs):
-    if (use_user_site is not None) and (not use_user_site):
-        logger.debug("Non-user install by explicit request")
-        return False
+    If use_user_site is None, the behavior is decided upon other
+    arguments and the environment.
 
-    if use_user_site:
-        if prefix_path:
-            raise CommandError(
-                "Can not combine '--user' and '--prefix' as they imply "
-                "different installation locations"
-            )
-        if virtualenv_no_global():
+    Compatibility of installation location options and the environment
+    is also asserted in this function.
+    """
+    # Check for incompatible options
+    locations = {"--target": target_dir, "--user": use_user_site,
+                 "--root": root_path, "--prefix": prefix_path}
+    destinations = [k for k, v in locations.items() if v]
+    if len(destinations) > 1:
+        last, rest = destinations[-1], ", ".join(destinations[:-1])
+        raise CommandError("Different installation locations are implied by "
+                           "{} and {}.".format(rest, last))
+
+    if target_dir:
+        if os.path.exists(target_dir) and not os.path.isdir(target_dir):
+            raise InstallationError("Target path exists but is not "
+                                    "a directory: {}".format(target_dir))
+        if not test_writable_dir(target_dir):
             raise InstallationError(
-                "Can not perform a '--user' install. User site-packages "
-                "are not visible in this virtualenv."
-            )
-        logger.debug("User install by explicit request")
-        return True
-
-    # If we are here, user installs have not been explicitly requested/avoided
-    assert use_user_site is None
-
-    # user install incompatible with --prefix/--target
-    if prefix_path or target_dir:
-        logger.debug("Non-user install due to --prefix or --target option")
+                "Cannot write to target directory: {}".format(target_dir))
+        logger.debug("Non-user install due to specified target directory")
         return False
-
-    # If user installs are not enabled, choose a non-user install
-    if not site.ENABLE_USER_SITE:
-        logger.debug("Non-user install because user site-packages disabled")
+    if prefix_path:
+        if not test_writable_dir(prefix_path):
+            raise InstallationError(
+                "Cannot write to prefix path: {}".format(prefix_path))
+        logger.debug("Non-user install due to specified prefix path")
         return False
 
-    # If we have permission for a non-user install, do that,
-    # otherwise do a user install.
-    if site_packages_writable(root=root_path, isolated=isolated_mode):
+    if use_user_site is not None:
+        use_user_site = bool(use_user_site)
+        logger.debug("{} install by explicit request".format(
+            "User" if use_user_site else "Non-user"))
+        if use_user_site is True and site.ENABLE_USER_SITE is None:
```

Comment:
`is True` is redundant. This var is always bool in this branch.
```suggestion
        if use_user_site and site.ENABLE_USER_SITE is None:
```

Reply:
Thank you for the review.  BTW is there any problem with the location conflict logic (since it will break backward compatibility I think)?  If it's OK then is it documented in the proper place (I appended it to `news/6762.bugfix`)?

Reply:
Well, if it introduces incompatible changes, it's probably a feature, not a bugfix.
Also, it's a question for @pradyunsg and other pip maintainers. I only did some basic coding review w/o any project-specific considerations mostly.